File: bindings/dart/lib/src/compaction_stats.dart
/// Statistics from a compaction run
class CompactionStats {
  final int filesBefore;
  final int filesAfter;
  final int entriesBefore;
  final int entriesAfter;
  final int tombstonesRemoved;
  final int sizeBefore;
  final int sizeAfter;

  CompactionStats({
    required this.filesBefore,
    required this.filesAfter,
    required this.entriesBefore,
    required this.entriesAfter,
    required this.tombstonesRemoved,
    required this.sizeBefore,
    required this.sizeAfter,
  });

  /// Bytes saved
  int get sizeReduction => (sizeBefore - sizeAfter).clamp(0, sizeBefore);

  /// Percentage of space saved
  double get sizeReductionPercent {
    if (sizeBefore == 0) return 0.0;
    return (sizeReduction / sizeBefore) * 100.0;
  }

  @override
  String toString() {
    return 'CompactionStats(files: $filesBefore→$filesAfter, '
        'tombstones: $tombstonesRemoved, '
        'reduction: ${sizeReductionPercent.toStringAsFixed(1)}%)';
  }
}

File: bindings/dotnet/FireLocal/CompactionStats.cs
namespace FireLocal
{
    /// <summary>
    /// Statistics from a compaction run
    /// </summary>
    public class CompactionStats
    {
        /// <summary>
        /// Number of SST files before compaction
        /// </summary>
        public int FilesBefore { get; set; }

        /// <summary>
        /// Number of SST files after compaction
        /// </summary>
        public int FilesAfter { get; set; }

        /// <summary>
        /// Number of entries before compaction
        /// </summary>
        public int EntriesBefore { get; set; }

        /// <summary>
        /// Number of entries after compaction
        /// </summary>
        public int EntriesAfter { get; set; }

        /// <summary>
        /// Number of tombstones removed
        /// </summary>
        public int TombstonesRemoved { get; set; }

        /// <summary>
        /// Total size in bytes before compaction
        /// </summary>
        public long SizeBefore { get; set; }

        /// <summary>
        /// Total size in bytes after compaction
        /// </summary>
        public long SizeAfter { get; set; }

        /// <summary>
        /// Bytes saved by compaction
        /// </summary>
        public long SizeReduction => Math.Max(0, SizeBefore - SizeAfter);

        /// <summary>
        /// Percentage of space saved
        /// </summary>
        public double SizeReductionPercent =>
            SizeBefore == 0 ? 0.0 : (SizeReduction / (double)SizeBefore) * 100.0;

        public override string ToString()
        {
            return $"CompactionStats(files: {FilesBefore}→{FilesAfter}, " +
                   $"tombstones: {TombstonesRemoved}, " +
                   $"reduction: {SizeReductionPercent:F1}%)";
        }
    }
}
